### PR TITLE
net/pidlisten: new package that restricts dials to the current process

### PR DESCRIPTION
**Pull request by crawshaw**
To be used in the C library wrapping tsnet to provide LocalAPI access.

This commit contains a linux implementation.
More operating systems to follow.

### CODE REVIEW

File: net/pidlisten/pidlisten_linux.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	"path/filepath"
+	"tailscale.com/util/dirwalk"
+
+	"github.com/vishvananda/netlink"
+)
+
+// NewPIDListener wraps a net.Listener so that it only accepts connections from the current process.
+func NewPIDListener(ln net.Listener) net.Listener {
+	return &listener{ln: ln}
+}
```
Comment (bradfitz) [inline net/pidlisten/pidlisten_linux.go:22]: I prefer the style of moving this to the main portable file (pidlisten.go) so the docs are in one place, and then having a `var newPIDListener func(net.Listener) net.Listener` that OS-specific files can init in their `init`.

Also: can this just be called `New`, given the package name?

File: net/pidlisten/pidlisten_linux.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	// that its inode belongs to the current PID.
+	s, err := netlink.SocketGet(conn.RemoteAddr(), conn.LocalAddr())
+	if err != nil {
+		return false, err
+	}
+
+	want := fmt.Sprintf("socket:[%d]", s.INode)
+	dir := fmt.Sprintf("/proc/%d/fd", os.Getpid())
+	err = dirwalk.WalkShallow(mem.S(dir), func(name mem.RO, de fs.DirEntry) error {
+		n, err := os.Readlink(filepath.Join(dir, name.StringCopy()))
```
Comment (bradfitz) [inline net/pidlisten/pidlisten_linux.go:53]: add TODO to remove this garbage like how portlist_linux.go does it with its readlink?  (maybe we export that somewhere)

File: net/pidlisten/pidlisten.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	}
+}
+
+func (pln *listener) Close() error {
+	return pln.ln.Close()
+}
+
+func (pln *listener) Addr() net.Addr {
+	return pln.ln.Addr()
+}
```
Comment (bradfitz) [inline net/pidlisten/pidlisten.go:42]: or embed `ln` instead and omit these. I don't care either way.

File: net/pidlisten/pidlisten_linux.go (hunk 1 above)
Comment (bradfitz) [reply net/pidlisten/pidlisten_linux.go:22]: Oh, this part is portable. You meant to move this to the other file.

File: net/pidlisten/pidlisten_noimpl.go [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+//go:build !linux
+// +build !linux
+
+package pidlisten
+
+import "net"
+
+func checkPIDLocal(conn net.Conn) (bool, error) {
+	panic("not implemented")
```
Comment (bradfitz) [inline net/pidlisten/pidlisten_noimpl.go:12]: just return an error for now? The signature permits it.

File: net/pidlisten/pidlisten_noimpl.go [hunk 5]
```diff
@@ -0,0 +1,13 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build !linux
+// +build !linux
```
Comment (bradfitz) [inline net/pidlisten/pidlisten_noimpl.go:5]: we removed +build from the tree already

File: net/pidlisten/pidlisten_test.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build linux
+// +build linux
```
Comment (bradfitz) [inline net/pidlisten/pidlisten_test.go:5]: delete +build

**Comment by crawshaw** [comment]
Planning to switch to a consistent http header approach that will work on iOS.

**Pull request closed without merging**